addons/controllers/calico/calicoconfig_controller.go

```diff
@@ -46,3 +88,181 @@ func (r *CalicoConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&cniv1alpha1.CalicoConfig{}).
 		Complete(r)
 }
+
+// ReconcileCalicoConfig reconciles CalicoConfig data values secrets
+func (r *CalicoConfigReconciler) ReconcileCalicoConfig(
+	ctx context.Context,
+	cluster *clusterapiv1beta1.Cluster,
+	calicoConfig *cniv1alpha1.CalicoConfig) (_ ctrl.Result, retErr error) {
+
+	var patchCRD bool
+
+	patchHelper, err := clusterapipatchutil.NewHelper(calicoConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// patch CalicoConfig before returning the function
+	defer func() {
+		// patchCRD will be true if finalizer or owner reference is added/removed
```

Comment:
question: will `patchCRD` be true on other cases as well? like CalicoConfig gets updated?

Reply:
It would be true only if finalizer or owner reference are added or removed for CalicoConfig. As of now we don't have logic for the modification of those fields for CalicoConfig.

Reply:
How about upgrade? Like a version bump on calico, there could be modifications on CalicoConfig ? Or maybe I am wrong. I am trying to think what are the cases we do not want to patch/update the enqueued `CalicoConfig` request. Few lines below, the logic seems always update the status.

Reply:
If CalicoConfig is getting updated, the finalizer and owner reference still remain the same, so the patch variable here will not be true and the ```patchHelper.Patch()``` won't get called. 
But for the Status().Update() we update it on Create/Update/Delete of CalicoConfig to refer to the reference to CalicoConfig or remove the reference. One might argue it is not required to update the status for when CalicoConfig is getting updated as the secret reference still remains the same, but I didn't want to make the logic more complicated by adding another check.

Reply:
If you look into the implementation of patchHelper.Patch(), it calculates the changes `before` and `after`, intelligently do the patch/update if there are changes and if the obj needs to be patched.  In addition to above, patchHelper.Patch() also patches status subresource.

My thinking is, if we could make thing simple: a) no need to update and check patchCRD bool b) just call patchHelper.Patch() on each reconciliation loop ?

Reply:
Removed ```patchCRD``` and now only calling ````patchHelper.Patch()``` on the caller ``ReconcileCalicoConfig()```